src/groupeddataframe/splitapplycombine.jl

```diff
@@ -1090,10 +1027,32 @@ function _agg2idx_map_helper(idx, idx_agg)
 end
 
 function _combine(f::AbstractVector{<:Pair},
-                  gd::GroupedDataFrame, nms::AbstractVector{Symbol})
+                  gd::GroupedDataFrame, nms::AbstractVector{Symbol},
+                  copycols::Bool, keeprows::Bool) # TODO: remove these defaults
     # here f should be normalized and in a form of source_cols => fun
     @assert all(x -> first(x) isa Union{Int, AbstractVector{Int}, AsTable}, f)
     @assert all(x -> last(x) isa Base.Callable, f)
+
+    if keeprows
+        if !_check_cannonical(gd)
+            throw(ArgumentError("select or transform functions require that " *
+                                "GroupedDataFrame is not subsetted"))
+        end
+        idx_keeprows = Vector{Int}(undef, nrow(parent(gd)))
+        let i = 0
+            for (s, e) in zip(gd.starts, gd.ends)
+                v = gd.idx[s]
+                for k in s:e
+                    i += 1
+                    idx_keeprows[i] = v
+                end
+            end
+            @assert i == nrow(parent(gd))
+        end
```

Comment:
```suggestion
        i = 0
        for (s, e) in zip(gd.starts, gd.ends)
            v = gd.idx[s]
            for k in s:e
                i += 1
                idx_keeprows[i] = v
            end
        end
        @assert i == nrow(parent(gd))
```

Reply:
This function is long and complex. I would prefer `i` to be defined only locally (it has no global meaning) - so that it does not pollute the namespace of `_combine`. Otherwise it is harder to debug cases if by accident `i` is used later before being properly initiated. Therefore I would leave the `let` block (although it does not improve anything performance-wise).

Reply:
I had to rewrite it for performance reasons anyway, so I resolve this comment.